Update sm doc (#415)
thin-edge/thin-edge.io: docs/src/tutorials/software-management.md

> # Manage your device software

Reviewer: ```suggestion
# Manage from Cumulocity the softwares install on your devices
```

Author: fixed

This document describes how to manage the software modules that are installed on a thin-edge device from the 
cloud using the **software management** feature of thin-edge.io.

> Note: As of now its supported only from Cumulocity cloud. Also, only supports **debian** based
 distributions(Ex: RaspberryPi OS , Ubuntu, Debian), which use the **apt** package manager.
 

> To enable software management feature on your device three components are required as mentioned below.

Reviewer: ```suggestion
Three components are required on your devices to enable software management:
```

Author: fixed

   Sm-mapper acts  as a proxy between the cloud and the device.
   This translates the cloud specific message type into device specific type and vice-versa.(Example: Cumulocity smart-rest to/from thin-edge json)
   The messages from cloud will be translated and forwarded to the `tedge_agent` and messages from `tedge_agent` will be translated and sent to cumulocity cloud.
   You can find this process with the name `tedge_mapper sm-c8y` in `ps` once it starts.

2. Software management agent
   The thin-edge software management agent is the one that calls the plugins.
   You can find this process with the name `tedge_agent` in `ps` once it starts.

3. Software management plugin
   Plugins are the interfaces that call the package manager (example: apt/apt-get) to do the software management operations (Install, Remove or update)
   You can find them in /etc/tedge/sm-plugins.
   As of now there is only one (apt) plugin is supported.

## Enable software management feature
 
Find more information about [how to install and enable software management.](../../howto-guides/0012_install_and_enable_software_management.md)
 
## Managing the device software on Cumulocity cloud

> Cumulocity cloud collects the references to the software for the thin-edge devices.

Reviewer: I don't understand that sentence.

Reviewer: I would say:

Managing the software packages installed on the devices of your Cumulocity tenant is a two steps operation.

1. Populate the **software repository** with all the software packages and versions you plan to install.
2. Trigger software update operations on the devices, to install specific packages of the repository.

Author: updated.

> # Manage from Cumulocity the softwares on your devices

Reviewer 2: ```suggestion
# Manage the softwares on your devices from Cumulocity
```

Author: updated

 distributions(Ex: RaspberryPi OS , Ubuntu, Debian), which use the **apt** package manager.

Three components are required on your devices to enable software management:

1. Software management mapper for Cumulocity cloud
   Sm-mapper acts  as a proxy between the cloud and the device.
   This translates the cloud specific message type into device specific type and vice-versa.(Example: Cumulocity smart-rest to/from thin-edge json)
   The messages from cloud will be translated and forwarded to the `tedge_agent` and messages from `tedge_agent` will be translated and sent to cumulocity cloud.
   You can find this process with the name `tedge_mapper sm-c8y` in `ps` once it starts.

2. Software management agent
   The thin-edge software management agent is the one that calls the plugins.
   You can find this process with the name `tedge_agent` in `ps` once it starts.

3. Software management plugin
   Plugins are the interfaces that call the package manager (example: apt/apt-get) to do the software management operations (Install, Remove or update)
   You can find them in /etc/tedge/sm-plugins.

>    As of now there is only one (apt) plugin is supported.

Reviewer 2: These 3 bullet points detailing the internal architecture: isn't this too much information for an end-user reading a tutorial? I think these process/service responsibility descriptions belong in a separate architecture/design section rather than here.

Author: Added as per @didier-wenzek suggestion. I feel as a dummy user, he/she should know what's happening inside the software management feature. So, felt its a good idea. Thats the difference between the howto vs tutorials.

Author: I put them in that order because it's easy to explain the operation flow from the cloud to the device.

## Enable software management feature
 
Find more information about [how to install and enable software management.](../../howto-guides/0012_install_and_enable_software_management.md)
 

> ## Managing the device software on Cumulocity cloud

Reviewer 2: ```suggestion
## Managing the device software from Cumulocity cloud
```

Author: "on cumulocity" is correct here. Because we are managing the device software on the cloud. It's about adding/deleting a software to repo.

Reviewer: Maybe the name repository is missing then?

Reviewer 2: Okay. In that case, how about "Managing the device software **repository** on Cumulocity cloud"?

> Managing the software packages installed on the devices of your Cumulocity tenant is a two steps operation.

Reviewer 2: ```suggestion
Managing the software packages installed on the devices from your Cumulocity tenant is a two steps operation.
```

Author: fixed

* Populate the software repository with all the software packages and versions you plan to install.

> * Trigger software update operations on the devices, to install specific packages of the repository. 

Reviewer 2: ```suggestion
* Trigger software update operations on the devices, to install specific packages from the repository. 
```

Author: fixed

2. Software management agent
   The thin-edge software management agent is the one that calls the plugins.
   You can find this process with the name `tedge_agent` in `ps` once it starts.

3. Software management plugin
   Plugins are the interfaces that call the package manager (example: apt/apt-get) to do the software management operations (Install, Remove or update)
   You can find them in /etc/tedge/sm-plugins.
   As of now there is only one (apt) plugin is supported.

## Enable software management feature
 
Find more information about [how to install and enable software management.](../../howto-guides/0012_install_and_enable_software_management.md)
 
## Managing the device software on Cumulocity cloud
Managing the software packages installed on the devices from your Cumulocity tenant is a two steps operation.
* Populate the software repository with all the software packages and versions you plan to install.
* Trigger software update operations on the devices, to install specific packages from the repository.

Find more information about [managing the device software](https://cumulocity.com/guides/users-guide/device-management/#managing-device-software)

> ### Adding new software into software repository

Reviewer: ```suggestion
### Adding new software into the software repository of your tenant
```

   - to add a new software, enter a name for the software (and confirm it by clicking Create new in the resulting window),
     a description and its version.
   - to add a new version, select the software for which you want to add a new version from the dropdown list in the Software

>      field and enter a version of the package. (The version is optional and can be left with a white space)

Reviewer: ```suggestion
     field and enter a version of the package. The version is optional and can be left with a white space, meaning the latest version.
```

Author: fixed

3. Optionally, you can define the device type filter when adding a new software.

> 4. As of now thin-edge.io only supports software from apt repositories. So, select the **Provide a file path** option and give an **empty space**.

Reviewer: I would add a note that the plugin type name that is provided here is used to pick the appropriate plugin among those installed in `/etc/tedge/sm-plugins`.

5. Click Save.

Follow above mentioned steps as shown in the diagram

 ![Add new software](./images/add-new-software-to-repo.png)

### Deleting software or software version
One can remove a software module or a specific version of it from the software repository.
Find more information about [how to delete the software or the specific software version](https://cumulocity.com/guides/users-guide/device-management/#deleting-softwares-or-software-versions)

> ## Managing software on a device

Reviewer: This section might be simplified with a list of bullet points.

```
From the Cumulocity cloud `Software` tab of a device, the software packages can be
* [installed](https://cumulocity.com/guides/users-guide/device-management/#to-install-software-on-a-device),
* [updated](https://cumulocity.com/guides/users-guide/device-management/#to-update-software-on-a-device)
*  or removed.
```

   - to add a new software, enter a name for the software (and confirm it by clicking Create new in the resulting window),
     a description and its version.
   - to add a new version, select the software for which you want to add a new version from the dropdown list in the Software
     field and enter a version of the package. The version is optional and can be left with a white space, meaning the latest version.

>    > Note: The version field format is `package_version::plugin_type_name`.The plugin type name that is provided here is used to pick

Reviewer 2: Wondering if this note is needed here until we support multiple plugins.

Author: Now this will not confuse the users if we even add a new plugin. Its to give more details to the user.

4. As of now thin-edge.io only supports software from apt repositories. So, select the **Provide a file path** option and give an **empty space**.
5. Click Save.

Follow above mentioned steps as shown in the diagram

 ![Add new software](./images/add-new-software-to-repo.png)

### Deleting software or software version
One can remove a software module or a specific version of it from the software repository.
Find more information about [how to delete the software or the specific software version](https://cumulocity.com/guides/users-guide/device-management/#deleting-softwares-or-software-versions)

## Managing software on a device
Find more information about [how to manage the software](https://cumulocity.com/guides/users-guide/device-management/#managing-software-on-a-device) on a device.

From the Cumulocity cloud `Software` tab of a device, software can be
* [installed](https://cumulocity.com/guides/users-guide/device-management/#to-install-software-on-a-device)
 > Note: Software profiles are not supported as of now on thin-edge.
* [updated](https://cumulocity.com/guides/users-guide/device-management/#to-update-software-on-a-device)

>  > Note: Thin-edge treats install and update same.

Reviewer 2: This note could have been added after the bullet points.

Author: I put there because its the right place. When user reads about the Update operation, s/he should get full details at one place.

> * removed

Reviewer 2: Provide this hyperlink: https://cumulocity.com/guides/users-guide/device-management/#to-delete-software-from-a-device for remove as well just like the other options?

>   Hover over the software entry which you want to remove and click the **Remove** software icon and click on `Apply changes` button.

Reviewer 2: We need one final sentence at the end that just says "click the `Apply changes` button" once the list of software to be installed/updated/removed is complete and wait for the execution status response from the device."

Author: Its already there in the install link, But I will add a note.

> ## Starting and Stopping the Software management services manually

Reviewer: I would expect this in the howto.

Reviewer: What would be useful here would be a hint on how to find on the device the cause of an error reported on the cloud.

Unfortunately, the UX is currently not good here.

Author: I agree with the first part. So, what to mention here?
